llm/server.go

```diff
@@ -214,15 +214,41 @@ func NewLlamaServer(gpus discover.GpuInfoList, model string, ggml *GGML, adapter
 		params = append(params, "--threads", strconv.Itoa(defaultThreads))
 	}
 
-	flashAttnEnabled := envconfig.FlashAttention()
+	fa := envconfig.FlashAttention()
+	if fa && !gpus.FlashAttentionSupported() {
+		slog.Warn("flash attention enabled but not supported by gpu")
+		fa = false
+	}
 
-	for _, g := range gpus {
-		// only cuda (compute capability 7+) and metal support flash attention
-		if g.Library != "metal" && (g.Library != "cuda" || g.DriverMajor < 7) {
-			flashAttnEnabled = false
+	if fa && !ggml.SupportsFlashAttention() {
+		slog.Warn("flash attention enabled but not supported by model")
+		fa = false
+	}
+
+	kvct := envconfig.KvCacheType()
+	if kvct == "" {
+		kvct = "f16"
+	}
```

Comment:
Leaving it as empty string _should_ fine here since it will simply not add the `--kv-cache-type` flag

Reply:
This was in response to some commented from @jessegross around making it f16 a more explicit default and thus doing away with all the extra checks for "" (and f16), can't find the specific github review comment, but mentioned here https://github.com/ollama/ollama/pull/6279#discussion_r1866826948

Reply:
I think it's safer for now to only add the flag if `kvct` is set (i.e. non empty string) since that avoids adding the `--kv-cache-type` flag below when the env var is not set. Should be a small change